src/RoboSaverVirtualModuleFactory.sol

```diff
@@ -0,0 +1,124 @@
+
+contract RoboSaverVirtualModuleFactory {
+    /*//////////////////////////////////////////////////////////////////////////
+                                     DATA TYPES
+    //////////////////////////////////////////////////////////////////////////*/
+    struct VirtualModuleDetails {
+        address virtualModuleAddress;
+        uint256 upkeepId;
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                   CONSTANTS
+    //////////////////////////////////////////////////////////////////////////*/
+
+    IERC20 constant LINK = IERC20(0xE2e73A1c69ecF83F464EFCE6A5be353a37cA09b2);
+    IKeeperRegistryMaster constant CL_REGISTRY = IKeeperRegistryMaster(0x299c92a219F61a82E91d2062A262f7157F155AC1);
+    IKeeperRegistrar constant CL_REGISTRAR = IKeeperRegistrar(0x0F7E163446AAb41DB5375AbdeE2c3eCC56D9aA32);
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                   PUBLIC STORAGE
+    //////////////////////////////////////////////////////////////////////////*/
+
+    // card -> (module address, upkeep id)
+    mapping(address => VirtualModuleDetails) public virtualModules;
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                       EVENTS
+    //////////////////////////////////////////////////////////////////////////*/
+    event RoboSaverVirtualModuleCreated(address virtualModule, address card, uint256 upkeepId, uint256 timestamp);
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                       ERRORS
+    //////////////////////////////////////////////////////////////////////////*/
+    error UpkeepZero();
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                     CONSTRUCTOR
+    //////////////////////////////////////////////////////////////////////////*/
+
+    constructor() {
+        LINK.approve(address(CL_REGISTRAR), type(uint256).max);
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                  EXTERNAL METHODS
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @notice Creates a virtual module for a card
+    /// @param _delayModule The address of the delay module
+    /// @param _rolesModule The address of the roles module
+    /// @param _buffer The buffer for the virtual module (configurable)
+    /// @param _slippage The slippage for the virtual module (configurable)
+    function createVirtualModule(address _delayModule, address _rolesModule, uint256 _buffer, uint16 _slippage)
+        external
+    {
+        // @todo sanity checks on the inputs!
+
+        // uses `CARD` address has to helps pre-determining the address of the virtual module given the salt
```

Comment:
thats a good point, in that case probably will throw due to the collision

assuming that a same CARD will deploy two different virtual modules we could use a counter to make the salt unique, but in which case do you envision a single CARD having two robo saver modules running in parallel? 🤔

Reply:
i guess if we were to support deploying newer versions from the same factory

alternative to that could be a new factory for every new version

Reply:
in the case that a new version will be supported by the factory itself, the `bytecode` will be different from old version, since its an element for the deterministic addy creation will not suffer collision

`new_address = hash(0xFF, sender, salt, bytecode)`

Reply:
yes that makes sense, this way there will be one unique module per user per version. sounds perfect

Reply:
@petrovska-petro still a suggestion here to commit